File: sach/de3300.py
//DE3300.PY - Tran Huu Nam - hsgtin.vn - 2024
import sys
from bisect import bisect_right
INF = 2e18
class DIEM:
    def __init__(self, a, b):
        self.a = a
        self.b = b
    def __lt__(self, other):
        return self.a < other.a
def main():
    input = sys.stdin.read
    data = input().split()
    idx = 0
    n = int(data[idx])
    m = int(data[idx + 1])
    c = int(data[idx + 2])
    idx += 3
    p = [None] * (n + 1)
    a = []
    for i in range(1, n + 1):
        p[i] = DIEM(int(data[idx]), int(data[idx + 1]))
        a.append(p[i].a)
        idx += 2
    p = sorted(p[1:], key=lambda x: x.a)
    a.sort()
    L = [INF] * (n + 1)
    L[0] = INF
    for i in range(1, n + 1):
        L[i] = min(L[i - 1], p[i - 1].b - p[i - 1].a * c)
    R = [INF] * (n + 2)
    R[n + 1] = INF
    for i in range(n, 0, -1):
        R[i] = min(R[i + 1], p[i - 1].b + p[i - 1].a * c)
    for i in range(m):
        d = int(data[idx])
        idx += 1
        k = bisect_right(a, d)
        answer = INF
        if k > 0:
            answer = min(answer, d * c + L[k])
        if k + 1 <= n:
            answer = min(answer, -d * c + R[k + 1])
        print(answer)
if __name__ == "__main__":
    main()

File: sach/de3300.cpp
//DE3300.CPP - Tran Huu Nam - hsgtin.vn - 2024
#include <bits/stdc++.h>
#define int long long
using namespace std;
const int INF = 2e18;
struct DIEM {
    int a, b;
    bool operator < (DIEM &other) {
        return a < other.a;
    }
};
signed main() {
    ios_base::sync_with_stdio(0); cin.tie(0);
    int n, m, c;
    cin >> n >> m >> c;
    vector<DIEM> p(n + 1);
    vector<int> a;
    for(int i = 1; i <= n; i ++) {
        cin >> p[i].a >> p[i].b;
        a.push_back(p[i].a);
    }
    sort(p.begin() + 1, p.begin() + n + 1);
    sort(a.begin(), a.end());
    vector<int> L(n + 1);
    L[0] = INF;
    for(int i = 1; i <= n; i ++)
        L[i] = min(L[i - 1], p[i].b - p[i].a * c);
    vector<int> R(n + 2);
    R[n + 1] = INF;
    for(int i = n; i >= 1; i --)
        R[i] = min(R[i + 1], p[i].b + p[i].a * c);
    for(int i = 1; i <= m; i ++) {
        int d;
        cin >> d;
        int k = upper_bound(a.begin(), a.end(), d)-a.begin();
        int answer = INF;
        if(k > 0) answer = min(answer, d * c + L[k]);
        if(k + 1 <= n) answer = min(answer, - d*c + R[k+1]);
        cout << answer << "\n";
    }
    return 0;
}

